src/routers/alpha-router/functions/get-candidate-pools.ts

```diff
@@ -941,5 +960,228 @@ export async function getV2CandidatePools({
     },
   };
 
-  return { poolAccessor, candidatePools: poolsBySelection };
+  return { poolAccessor, candidatePools: poolsBySelection, subgraphPools };
+}
+
+export async function getMixedRouteCandidatePools({
+  tokenIn,
+  tokenOut,
+  routeType,
+  routingConfig,
+  V3subgraphProvider,
+  V2subgraphProvider,
+  tokenProvider,
+  V3poolProvider,
+  V2poolProvider,
+  blockedTokenListProvider,
+  chainId,
+}: MixedRouteGetCandidatePoolsParams): Promise<{
+  V2poolAccessor: V2PoolAccessor;
+  V3poolAccessor: V3PoolAccessor;
+  candidatePools: CandidatePoolsBySelectionCriteria;
+  subgraphPools: (V2SubgraphPool | V3SubgraphPool)[];
+}> {
+  const { blockNumber } = routingConfig;
+  const { subgraphPools: V3subgraphPools, candidatePools: V3candidatePools } =
+    await getV3CandidatePools({
+      tokenIn,
+      tokenOut,
+      tokenProvider,
+      blockedTokenListProvider,
+      poolProvider: V3poolProvider,
+      routeType,
+      subgraphProvider: V3subgraphProvider,
+      routingConfig,
+      chainId,
+    });
+  const { subgraphPools: V2subgraphPools, candidatePools: V2candidatePools } =
+    await getV2CandidatePools({
+      tokenIn,
+      tokenOut,
+      tokenProvider,
+      blockedTokenListProvider,
+      poolProvider: V2poolProvider,
+      routeType,
+      subgraphProvider: V2subgraphProvider,
+      routingConfig,
+      chainId,
+    });
+
+  /**
+   * Implement heuristic: picking V2 pools with higher liq than respective V3 pools, or if v3 pool doesn't exist
+   */
+  /// We only really care about pools involving the tokenIn or tokenOut explictly,
+  /// since there's no way a long tail token in V2 would be routed through as an intermediary
+  const V2topByTVLPools = [
+    ...V2candidatePools.selections.topByTVLUsingTokenIn,
+    ...V2candidatePools.selections.topByBaseWithTokenIn,
+    /// tokenOut:
+    ...V2candidatePools.selections.topByTVLUsingTokenOut,
+    ...V2candidatePools.selections.topByBaseWithTokenOut,
+    /// Direct swap:
+    ...V2candidatePools.selections.topByDirectSwapPool,
+  ];
+
+  const V2topByTVLSortedPools = _(V2subgraphPools)
+    .filter((pool) => V2topByTVLPools.map((p) => p.id).includes(pool.id))
+    .sortBy((pool) => -pool.reserveUSD)
+    .value();
+
+  /// we consider all returned V3 pools for this heuristic to "fill in the gaps"
+  const V3sortedPools = _(V3subgraphPools)
+    .sortBy((pool) => -pool.tvlUSD)
+    .value();
+
+  let buildV2Pools: V2SubgraphPool[] = [];
```

Comment:
Yea that thought crossed my mind, however from what I've seen in the results from this heuristic, we add V2 pools to consideration because there is no liquidity on V3 for that token pair much more often than because of them having more USD in the pair. I don't think it's that bad to be a little more loose with selecting V2 pools and tossing them out at the end if they're not favorable since we're not considering too many additional routes rn with `mixedRoutes` being added.

Reply:
I also have no idea how we'd compare liq depth based on price ranges lol, we would likely have to get tick data and then we might as well just calculate the quote off chain